sonar-scm-git-plugin/src/main/java/org/sonarsource/scm/git/FindLatestForkWalk.java

```diff
@@ -0,0 +1,178 @@
+
+        if (origins.size() > 1 && origins.contains(currentBranch)) {
+          return matchFound(next, origins);
+        }
+
+        for (RevCommit p : next.getParents()) {
+          if (p.equals(currentBranch.head)) {
+            // we are not interested in branches where the current branch is part of
+            continue;
+          }
+
+          // carry origins to parents, even if it has been seen before
+          registerOrigins(p, origins);
+
+          if (!p.has(RevFlag.SEEN)) {
+            // parse now to have commit time for the queue
+            parse(walk, p);
+            p.add(RevFlag.SEEN);
+            queue.add(p);
+          }
+        }
+
+        commitOrigins.remove(next);
+      }
+    }
+
+    return null;
+  }
+
+  private static ForkPoint matchFound(RevCommit commit, List<Branch> origins) {
+    Set<String> references = origins.stream()
+      .map(b -> b.name)
+      .filter(n -> !"HEAD".equals(n))
+      .map(n -> n.startsWith("refs/heads/") ? n.substring(11) : n)
+      .collect(Collectors.toSet());
+    return new ForkPoint(commit.getName(), references);
+  }
+
+  private static class Branch {
+    private final RevCommit head;
+    private final String name;
+
+    private Branch(RevCommit head, String name) {
+      this.head = head;
+      this.name = name;
+    }
+  }
+
+  private Optional<RevCommit> refToCommit(RevWalk walk, Ref ref) throws IOException {
+    RevObject maybehead = walk.parseAny(ref.getObjectId());
+    if (!(maybehead instanceof RevCommit)) {
+      return Optional.empty();
+    }
+    return Optional.of((RevCommit) maybehead);
+  }
+
+  private Set<Branch> allRefsExceptCurrent(RevWalk walk, ObjectId currentObjectId) throws IOException {
+    Set<Branch> branches = new HashSet<>();
+
+    for (Ref ref : repo.getRefDatabase().getRefs()) {
+      if (ref.getObjectId().equals(currentObjectId) || "HEAD".equals(ref.getName())) {
+        continue;
+      }
+
+      Optional<RevCommit> commit = refToCommit(walk, ref);
+      commit.ifPresent(c -> branches.add(new Branch(c, ref.getName())));
+    }
+
+    return branches;
+  }
+
+  private void init(RevWalk walk, Branch current, Set<Branch> otherRefs) throws IOException {
+    registerOrigin(current.head, current);
+    current.head.add(RevFlag.SEEN);
+    parse(walk, current.head);
+    queue.add(current.head);
+
+    for (Branch c : otherRefs) {
+      parse(walk, c.head);
+      registerOrigin(c.head, c);
+      if (!c.head.has(RevFlag.SEEN)) {
+        // we might have 2 references pointing to the same commit
+        c.head.add(RevFlag.SEEN);
+        queue.add(c.head);
+      }
+    }
+  }
+
+  /**
+   * Parse commit's headers so that we have access to its parents and commit time
+   */
+  private void parse(RevWalk walk, RevCommit c) throws IOException {
+    walk.parseHeaders(c);
+  }
+
+  private void registerOrigin(RevCommit commit, Branch origin) {
+    if (origin == null) {
```

Comment:
it shouldn't be null, this is to detect early a problem if it is. Do you think marking it as Nullable would be better? I think it would document it in the wrong way

Reply:
We already define `@ParametersAreNonnullByDefault` on the package in this project just as in `sonar-enterprise`, so it seems to me if the param is not `@Nullable` we shouldn't have code handling it (especially since this is a private method). This doesn't mean that someone won't attempt to pass null but IDE should warn him about potential issue. Not sure which IDE you use, perhaps it's not universally supported yet? You definitely have warnings in IntelliJ.

Or maybe I'm misunderstanding how this should be used. @pierre-guillot-sonarsource @henryju any opinions?

Reply:
Talking about sonar-enterprise, we extensively use `Preconditions.checkNonNull` in it. To clarify this parameter is not nullable and shouldn't be documented as so (it will throw an ISE afterall).

Reply:
There is also `Objects.requireNonNull` in JDK